core/src/main/kotlin/org/eln2/data/ComponentGraph.kt

```diff
@@ -0,0 +1,278 @@
+package org.eln2.data
+
+import org.eln2.debug.dprintln
+import java.lang.StringBuilder
+import java.util.*
+
+open class Edge(val a: Vertex, val b: Vertex): Iterable<Vertex> {
+	override fun iterator() = arrayOf(a, b).iterator()
+}
+
+open class Vertex(val graph: ComponentGraph) {
+	var component = graph.constructComponent(this)
+	val incident: MutableMap<Vertex, Edge> = mutableMapOf()
+}
+
+open class ConnectedComponent(initialVertex: Vertex) {
+	val vertices: MutableSet<Vertex> = mutableSetOf(initialVertex)
+	
+	internal open fun add(v: Vertex) = vertices.add(v)
+	internal open fun remove(v: Vertex) = vertices.remove(v)
+	internal open fun clear() = vertices.clear()
+
+	internal var realized = false
+	internal open fun realize() {}
+	internal open fun unrealize() {}
+}
+
+interface ComponentListener {
+	fun componentRealized(c: ConnectedComponent)
+	fun componentUnrealized(c: ConnectedComponent)
+}
+
+interface VertexListener {
+	fun vertexAdded(v: Vertex)
+	fun vertexRemoved(v: Vertex)
+}
+
+interface EdgeListener {
+	fun edgeAdded(e: Edge)
+	fun edgeRemoved(e: Edge)
+}
+
+open class ComponentGraph {
+	val vertices: MutableSet<Vertex> = mutableSetOf()
+	val edges: MutableSet<Edge> = mutableSetOf()
+	var components: MutableSet<ConnectedComponent> = mutableSetOf()
+	
+	val componentListeners: MutableSet<ComponentListener> = mutableSetOf()
+	val vertexListeners: MutableSet<VertexListener> = mutableSetOf()
+	val edgeListeners: MutableSet<EdgeListener> = mutableSetOf()
+	
+	fun addListener(listener: Any) {
+		if(listener is ComponentListener) componentListeners.add(listener)
+		if(listener is VertexListener) vertexListeners.add(listener)
+		if(listener is EdgeListener) edgeListeners.add(listener)
+	}
+	
+	fun removeListener(listener: Any) {
+		componentListeners.remove(listener)
+		vertexListeners.remove(listener)
+		edgeListeners.remove(listener)
+	}
+
+	internal open fun constructComponent(v: Vertex) = ConnectedComponent(v)
+	internal open fun constructVertex() = Vertex(this)
+	internal open fun constructEdge(a: Vertex, b: Vertex) = Edge(a, b)
+	
+	internal fun newVertex() = constructVertex().apply {
+		vertices.add(this)
+		components.add(component)
+		vertexListeners.forEach { it.vertexAdded(this) }
+	}
+	
+	internal fun removeVertex(v: Vertex) {
+		vertexListeners.forEach { it.vertexRemoved(v) }
+		disconnect(v)
+		vertices.remove(v)
+	}
+
+	internal fun connect(a: Vertex, b: Vertex): Edge {
+		assert(a.graph == this) { "CG.c: vertex a=$a not in graph $this" }
+		assert(b.graph == this) { "CG.c: vertex b=$b not in graph $this" }
```

Comment:
I think we have cleaner debug prints but perhaps it would be worth having a `dsf()` (debug string format) call to do this cleanly and consistently.